src/bios.rs

```diff
@@ -1,627 +1,625 @@
-  }
-}
+//! * All BIOS functions clobber `r0`, `r1`, and `r3`.
+//! * Some functions also use `r2` as an input register.
+//! * All other registers are unaffected.
 
-/// (`swi 0x01`) RegisterRamReset.
-///
-/// Clears the portions of memory given by the `flags` value, sets the Display
-/// Control Register to `0x80` (forced blank and nothing else), then returns.
-///
-/// * Flag bits:
-///   0) Clears the 256k of EWRAM (don't use if this is where your function call
-///      will return to!)
-///   1) Clears the 32k of IWRAM _excluding_ the last `0x200` bytes (see also:
-///      the `soft_reset` function)
-///   2) Clears all Palette data
-///   3) Clears all VRAM
-///   4) Clears all OAM (reminder: a zeroed object isn't disabled!)
-///   5) Reset SIO registers (resets them to general purpose mode)
-///   6) Reset Sound registers
-///   7) Reset all IO registers _other than_ SIO and Sound
-///
-/// **Bug:** The LSB of `SIODATA32` is always zeroed, even if bit 5 was not
-/// enabled. This is sadly a bug in the design of the GBA itself.
-///
-/// ## Safety
-///
-/// It is generally a safe operation to suddenly clear any part of the GBA's
-/// memory, except in the case that you were executing out of EWRAM and clear
-/// that. If you do then you return to nothing and have a bad time.
-#[inline(always)]
-pub unsafe fn register_ram_reset(flags: RegisterRAMResetFlags) {
-  #[cfg(not(target_arch = "arm"))]
-  {
-    unimplemented!("This function is not supported on this target.")
-  }
-  #[cfg(target_arch = "arm")]
-  {
-    asm!("swi 0x01", in("r0") flags.0);
-  }
+// Note(Lokathor): This makes intra-doc links work.
+#[allow(unused)]
+use crate::prelude::*;
+
+/// (`swi 0x00`) Performs a "soft reset" of the device.
+///
+/// Loads `r14` based on the `u8` value at `0x0300_7FFA`:
+/// * zero: `0x0800_0000` (ROM)
+/// * non-zero: `0x0200_0000` (EWRAM)
+///
+/// Then resets the following memory and registers:
+/// * `0x300_7E00` ..= `0x300_7FFF`: zeroed
+/// * `r0` ..= `r12`: zeroed
+/// * `sp_usr`: `0x300_7F00`
+/// * `sp_irq`: `0x300_7FA0`
+/// * `sp_svc`: `0x300_7FE0`
+/// * `lr_svc`, `lr_irq` : zeroed
+/// * `spsr_svc`, `spsr_irq`: zeroed
+///
+/// Then jumps to the `r14` address. This never returns.
+pub unsafe fn SoftReset() -> ! {
+  asm!("swi 0x00", options(noreturn))
 }
-newtype! {
-  /// Flags for use with `register_ram_reset`.
-  RegisterRAMResetFlags, u8
+
+/// (`swi 0x01`) Resets RAM and/or IO registers
+///
+/// * Note that if the IWRAM flag is used it doesn't reset the final `0x200`
+///   bytes of IWRAM. Instead, those bytes are reset during a call to the
+///   [`SoftReset`] function.
+/// * BIOS Bug: Data in `SIODATA32` is always destroyed, even if the `sio` flag
+///   is not set.
+#[inline]
+#[instruction_set(arm::t32)]
+pub unsafe fn RegisterRamReset(flags: crate::mmio_types::ResetFlags) {
+  asm!("swi 0x01",
+    inlateout("r0") flags.0 => _,
+    out("r1") _,
+    out("r3") _,
+    options(nomem, nostack)
```

Comment:
RegisterRamReset is marked 'nomem'. I'm not sure this is correct.

[The inline asm! RFC](https://github.com/Amanieu/rfcs/blob/inline-asm/text/0000-inline-asm.md) states:

> nomem means that the asm code does not read or write to memory. By default the compiler will assume that inline assembly can read or write any memory address that is accessible to it (e.g. through a pointer passed as an operand, or a global).

RegisterRamReset's purpose is to clear out RAM and the I/O registers. For the purpose of compilation and optimization, we must consider this like writing a global the size of (nearly) all RAM.

Reply:
Yeah, this one needs more documentation. Basically you can't actually reset iwram ever without immediately breaking rust's rules. You also can't reset ewram if you're using it as normal memory.

however you *can* reset any volatile-only memory regions without it being counted against using the `nomem` modifier.

we probably/maybe want a compiler fence on this one too.

this one is a pain.

Reply:
The only volatile-only memory is the I/O block (and palette/vram/oam if rust and/or the programmer pinky promises never to put variables there)

Reply:
Rust will not on its own start using any part of VRAM (on your own you just get stack variables, which are in iwram of course).

So the only people who would are either going through the safe interface from this crate... which I have not fully written yet but which will be largely volatile and won't allow references into vram, which is what could cause the problem. Rust references have rather strict rules, and they're not really a goof fit for vram.